thunder/torch/__init__.py

```diff
@@ -5331,6 +5336,223 @@ def wait(slf) -> None:
+    elif isinstance(inp, torch.dtype):
+        return _torch_to_thunder_dtype_map[inp]
+    elif isinstance(inp, (str, bool, int, float, complex)):
+        return inp
+    else:
+        raise NotImplementedError(f"Unsupported type: {builtins.type(inp)}")
+
+
+def augmented_forward_adaptor(sym_op: Callable):
+    def wrapper(*args, **kwargs):
+        from thunder.core.transforms import VJPDual
+
+        out = sym_op(*args, **kwargs)
+        primal = out if isinstance(out, tuple) else (out,)
+        residuals = ((args, kwargs),)
+        return VJPDual(primal, residuals)
+
+    return wrapper
+
+
+def backward_adaptor(torch_func: Callable):
+    def wrapper(saved_for_backward, *grad_output):
+        inp_args, inp_kwargs = saved_for_backward
+        flat_args, spec = tree_flatten(inp_args)
+        if not any(map(_is_differentiable, grad_output)):
+            return tree_unflatten(len(flat_args) * [None], spec)
+        if any(map(_is_differentiable, tree_flatten(inp_kwargs)[0])):
+            raise NotImplementedError(
+                f"Exception encountered when doing automatic registration for {torch_func} because there is keyword argument that requires gradient, please use manual registration."
+            )
+        from torch._subclasses.fake_tensor import FakeTensorMode
+
+        with FakeTensorMode() as mode:
+            fake_inp_args, fake_inp_kwargs = tree_map(lambda x: _get_fake_arg(x, mode), (inp_args, inp_kwargs))
+            wrapped_torch_func, diff_args = _make_differentiable_wrapper(torch_func, *fake_inp_args, **fake_inp_kwargs)
+            try:
+                _, fake_vjp = torch.autograd.functional.vjp(
+                    wrapped_torch_func, diff_args, v=tree_map(lambda x: _get_fake_arg(x, mode), grad_output)
+                )
+            except Exception as e:
+                msg = f"Exception encountered when doing automatic registration for {torch_func}, please use manual registration: {repr(e)}"
+                raise NotImplementedError(msg) from e
+
+        out_vjp = tree_map(_fake_type_to_thunder, fake_vjp)
+        iter_out_vjp = iter(out_vjp if isinstance(out_vjp, Sequence) else (out_vjp,))
+        fake_vjp = tuple(next(iter_out_vjp) if _is_differentiable(arg) else None for arg in flat_args)
+        return tree_unflatten(fake_vjp, spec)
+
+    return wrapper
+
+
+def _vjp_impl(torchfn, residules, *gs):
+    inp_args, inp_kwargs = residules
+
+    wrapped_func, diff_args = _make_differentiable_wrapper(torchfn, *inp_args, **inp_kwargs)
+    _, vjp_outs = torch.autograd.functional.vjp(wrapped_func, diff_args, v=gs)
+    flat_inp_args, inp_spec = tree_flatten(inp_args)
+    iter_vjp_out = iter(vjp_outs if isinstance(vjp_outs, Sequence) else (vjp_outs,))
+    vjp_outs = tuple(next(iter_vjp_out) if _is_differentiable(arg) else None for arg in flat_inp_args)
+    return tree_unflatten(vjp_outs, inp_spec)
+
+
+def meta_adaptor(torch_func: Callable):
+    def wrapper(*args, **kwargs):
```

Comment:
It's not a wrapper because this function doesn't wrap `torch_func` it constructs a meta function for `torch_func`.
```suggestion
    def meta_func(*args, **kwargs):
```